contributors/design-proposals/service_links.md

```diff
@@ -0,0 +1,254 @@
+
+When a service referenced using this API does not exist, the Kubelet should
+avoid starting the pod.
+
+## Proposed Changes
+
+### API changes
+
+#### Expressing dependencies on services
+
+```go
+type PodSpec {
+  // other fields omitted
+
+  // ServiceDependencies is a list of services in this pod's namespace that the
+  // pod depends on in order to start.  
+  ServiceDependencies []LocalObjectReference `json:"serviceDependencies,omitempty"`
+}
+```
+
+#### Environment variable injection
+
+Pull request [#148](https://github.com/kubernetes/community/pull/148)
+proposes an `EnvFrom` addition to the PodSpec to handle creating multiple
+environment variables from a single subject.  This API surface is ideal for the
+service use-cases, so we will extend it here:
+
+```go
+type Container struct {
+	// other fields omitted
+
+	EnvFrom []EnvFromSource `json:"envFrom,omitempty"`
+}
+
+type EnvVarFromSource struct {
+	// other fields omitted
+
+	// Prefix is an optional prefix for generated environment variables
+	Prefix string `json:"prefix,omitempty"`
+
+	// Can be refactored to ObjectReference at a later time while
+	// maintaining backward compatibility.
+	Service *LocalObjectReference `json:"service,omitempty"`
+}
+```
+
+When a pod uses this API feature, it indicates that the automatic injection
+behavior is not desired for that pod.  No automatic injection will occur when
+any container in a pod uses an explicit link to a service.
+
+Note, the `kubernetes` and `kubernetes-ro` services will always be injected,
+regardless of the state of the pod's spec.
+
+### Kubelet Changes
+
+The Kubelet must be modified to:
+
+1. Support creation of environment variables via the new API
+2. Disable the automatic injection for a pod whenever an explicit service link
+   appears in the pod's spec
+3. Start the containers in a pod only when all of the pod's explicitly linked
+   services exist
+4. Create an event for the pod when pod start is delayed due to waiting for a
+   service to exist
+
+## Examples
+
+### No custom variable names
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: simple-service-reference-example
+spec:
+  containers:
+  - image: busybox
+    name: example-container
+    envFrom:
+    - service:
+        name: redis-master
+```
+
+Yields the environment variables:
+
+```
+REDIS_MASTER_SERVICE_HOST=10.0.0.11
+REDIS_MASTER_SERVICE_PORT=6379
+REDIS_MASTER_PORT=tcp://10.0.0.11:6379
+REDIS_MASTER_PORT_6379_TCP=tcp://10.0.0.11:6379
+REDIS_MASTER_PORT_6379_TCP_PROTO=tcp
+REDIS_MASTER_PORT_6379_TCP_PORT=6379
+REDIS_MASTER_PORT_6379_TCP_ADDR=10.0.0.11
+```
+
+### Custom variable names
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: simple-service-reference-example
+spec:
+  containers:
+  - image: busybox
+    name: example-container
+    envFrom:
+    - prefix: "CUSTOM_NAME"
+      service:
+        name: redis-master
+```
+
+Yields the environment variables:
+
+```
+CUSTOM_NAME_SERVICE_HOST=10.0.0.11
```

Comment:
Given your two examples, the one with a prefix should have a result like:
CUSTOM_NAMEREDIS_MASTER_SERVICE_HOST=10.0.0.11
CUSTOM_NAMEREDIS_MASTER_SERVICE_PORT=6379

Not very readable but it is how it works.